File: DAIDALUS/Java/src/gov/nasa/larcfm/ACCoRD/TrafficState.java
/*
 * Copyright (c) 2015 United States Government as represented by
 * the National Aeronautics and Space Administration.  No copyright
 * is claimed in the United States under Title 17, U.S.Code. All Other
 * Rights Reserved.
 */
package gov.nasa.larcfm.ACCoRD;

import java.util.List;

import gov.nasa.larcfm.Util.Position;
import gov.nasa.larcfm.Util.Vect3;
import gov.nasa.larcfm.Util.Velocity;

public class TrafficState {
  protected final String id;
  protected final Position pos;
  protected final Velocity vel;

  public TrafficState() {
    id = "_NoAc_";
    pos = Position.INVALID;
    vel = Velocity.INVALID;
  }

  public static final TrafficState INVALID = new TrafficState();

  public TrafficState(String id, Position pos, Velocity vel) {
    this.id = id;
    this.pos = pos;
    this.vel = vel;
  }

  public TrafficState(TrafficState ac) {
    this(ac.id,ac.pos,ac.vel);
  }

  public boolean isValid() {
    return !pos.isInvalid() && !vel.isInvalid();
  }

  public String getId() {
    return id;
  }

  public boolean isLatLon() {
    return pos.isLatLon();
  }

  public Position getPosition() {
    return pos;
  }

  public Velocity getVelocity() {
    return vel;
  }

  public TrafficState linearProjection(double offset) {
    return new TrafficState(id, pos.linear(vel, offset), vel);
  }

  public boolean sameId(TrafficState ac) {
    return isValid() && ac.isValid() && id.equals(ac.id);
  }

  public static TrafficState getTraffic(List<TrafficState> traffic, String id) {
    if (!id.equals(TrafficState.INVALID.getId())) {
      for (int i=0; i < traffic.size(); ++i) {
        TrafficState ac = traffic.get(i);
        if (ac.getId().equals(id))
          return ac;
      }
    }
    return TrafficState.INVALID;
  }
  
  public String toString() {
    return "("+id+", "+pos.toString()+", "+vel.toString()+")";
  }

  public static String toPVS(String id, Vect3 s, Velocity v, int prec) {
    return "(# id := \"" + id + "\", s := "+s.toPVS(prec)+", v := "+v.toPVS(prec)+" #)";   
  }

}

File: DAIDALUS/C++/src/TrafficState.cpp
/*
 * Copyright (c) 2015 United States Government as represented by
 * the National Aeronautics and Space Administration.  No copyright
 * is claimed in the United States under Title 17, U.S.Code. All Other
 * Rights Reserved.
 */
#include "TrafficState.h"
#include <string>
#include "Position.h"
#include "Velocity.h"
#include "string_util.h"

namespace larcfm {

TrafficState::TrafficState() {
  id = "_NoAc_";
  pos = Position::INVALID();
  vel = Velocity::INVALIDV();
}

TrafficState::TrafficState(const std::string& i, const Position& p, const Velocity& v) {
  id = i;
  pos = p;
  vel = v;
}

TrafficState::TrafficState(const TrafficState& ac) {
  id = ac.id;
  pos = ac.pos;
  vel = ac.vel;
}

const TrafficState TrafficState::INVALID = TrafficState();

bool TrafficState::isValid() const {
  return !pos.isInvalid() && !vel.isInvalid();
}

std::string TrafficState::getId() const {
  return id;
}

bool TrafficState::isLatLon() const {
  return pos.isLatLon();
}

Position TrafficState::getPosition() const {
  return pos;
}

Velocity TrafficState::getVelocity() const {
  return vel;
}

TrafficState TrafficState::linearProjection(double offset) const {
  return TrafficState(id, pos.linear(vel,offset), vel);
}

bool TrafficState::sameId(const TrafficState& ac) const {
  return isValid() && ac.isValid() && id == ac.id;
}

std::string TrafficState::toString() const {
  return "("+id+", "+pos.toString()+", "+vel.toString()+")";
}

std::string TrafficState::toPVS(const std::string& id, const Vect3& s, const Velocity& v, int prec) {
  return "(# id := \"" + id + "\", s := "+s.toPVS(prec)+", v := "+v.toPVS(prec)+" #)";
}


std::string TrafficState::FmAircraftList(const std::vector<TrafficState>& traffic) {
  std::string s = "[";
  bool comma = false;
  for (int i=0; i < traffic.size(); ++i) {
    if (comma) {
      s += ", ";
    } else {
      comma = true;
    }
    s += traffic[i].toString();
  }
  return s+"]";
}

TrafficState TrafficState::getTraffic(const std::vector<TrafficState>& traffic, const std::string& id) {
  if (id != TrafficState::INVALID.getId()) {
    for (int i=0; i < (int) traffic.size(); ++i) {
      TrafficState ac = traffic[i];
      if (id == ac.getId())
        return ac;
    }
  }
  return TrafficState::INVALID;
}

// global scope:

bool operator<( const TrafficState& lhs , const TrafficState& rhs )
{
  return lhs.getId() < rhs.getId();
}

}
